rust_src/src/lisp.rs

```diff
@@ -147,19 +176,206 @@ pub enum LispMiscType {
     Lisp_Misc_Finalizer,
 }
 
-// Number of bits in a Lisp_Object tag.
-const GCTYPEBITS: libc::c_int = 3;
 
-const INTTYPEBITS: libc::c_int = GCTYPEBITS - 1;
+// Lisp_Misc is a union. Now we don't really care about its variants except the
+// super type layout. LispMisc is an unsized type for this, and LispMiscAny is 
+// only the header and a padding, which is consistent with the c version.
+// directly creating and moving or copying this struct is simply wrong!
+// If needed, we can calculate all variants size and allocate properly.
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+pub struct ExternalPtr<T>(* mut T);
+
+impl<T> ExternalPtr<T> {
+    pub fn new(p: *mut T) -> ExternalPtr<T> {
+        ExternalPtr(p)
+    }
+
+    pub fn as_ptr(&self) -> * const T {
+        self.0
+    }
+}
+
+impl<T> Deref for ExternalPtr<T> {
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        unsafe { &*self.0 }
+    }
+}
+
+pub type LispMiscRef = ExternalPtr<LispMiscAny>;
+
+// Supertype of all Misc types.
+#[repr(C)]
+pub struct LispMiscAny {
+    pub ty: LispMiscType,
+    // This is actually a GC marker bit plus 15 bits of padding, but
+    // we don't care right now.
+    padding: u16,
+}
+
+#[test]
+fn test_lisp_misc_any_size() {
+    // Should be 32 bits, which is 4 bytes.
+    assert!(mem::size_of::<LispMiscAny>() == 4);
+}
+
+impl LispObject {
+    pub fn is_misc(self) -> bool {
+        unsafe { self.get_type_unchecked() == LispType::Lisp_Misc }
+    }
+
+    pub unsafe fn to_misc_unchecked(self) -> LispMiscRef {
+        LispMiscRef::new(mem::transmute(self.get_untaggedptr_unchecked()))
+    }
+
+    pub unsafe fn get_misc_type_unchecked(self) -> LispMiscType {
+        self.to_misc_unchecked().ty
+    }
+}
+
+// Integer support (LispType == Lisp_Int0 | Lisp_Int1 == 2 | 6(LSB) )
 
 // Largest and smallest numbers that can be represented as integers in
 // Emacs lisp.
-const MOST_POSITIVE_FIXNUM: EmacsInt = EMACS_INT_MAX >> INTTYPEBITS;
+pub const MOST_POSITIVE_FIXNUM: EmacsInt = EMACS_INT_MAX >> INTTYPEBITS;
 #[allow(dead_code)]
-const MOST_NEGATIVE_FIXNUM: EmacsInt = (-1 - MOST_POSITIVE_FIXNUM);
+pub const MOST_NEGATIVE_FIXNUM: EmacsInt = (-1 - MOST_POSITIVE_FIXNUM);
 
-// This is also dependent on USE_LSB_TAG, which we're assuming to be 1.
-const VALMASK: EmacsInt = -(1 << GCTYPEBITS);
+
+// Lisp_Int0 (natnum)
+
+/// Natnums are inline integers that is great or equal to 0 and fit directly into
```

Comment:
The sentence alone is correct (natnum = natural number) but it didn't explain the whole thing correct. And natnum doesn't have to be in Lisp_Int0 in LSB mode. I was wrong here :(

I'll add definition for fixnum and fix this comment. fixnum operations will be in TODO.
(The original code is limiting and confusing... maybe because of the LSB and MSB difference.
Might fix this later when more things are settled)